### DISCUSSION TITLE
Towards less abuse of `asm`

### DISCUSSION THREAD

**Discussion by gatling-nrl** [kinnala/scikit-fem#854] [Ideas]
Ok, so yeah, I like to inflict syntactic mischief with `Functional` and `asm`. Abuse, some would say. [_Have said_](https://github.com/kinnala/scikit-fem/discussions/824#discussion-3724464), in fact.

But its just so _useful_ for poking around at things...

Is there any appetite for an actual iterator that traverses the data structures similar to how I've been using `asm`?

This is a wide open question... obviously you can get quite far with what exists already. But I admit it's side effects, and it's ugly... And I'm not too clear at all on syntax or anything... its just an idea.

The first use case I was thinking of is all this plotting and poking around I've been doing. But then you can look at my train wreck in https://github.com/kinnala/scikit-fem/pull/851 and see how maybe it could be quite useful when trying to extend skfem.

**Comment by gdmcbain** [comment]
> Abuse, some would say. [_Have said_](https://github.com/kinnala/scikit-fem/discussions/824#discussion-3724464), in fact.

Ha, yes.  Just in case this should be taken wrongly, [&lsquo;abuse of notation&rsquo;](https://en.wikipedia.org/wiki/Abuse_of_notation) is a common locution in metamathematics and not at all derogatory:

> In mathematics, **abuse of notation** occurs when an author uses a mathematical notation in a way that is not entirely formally correct, but which might help simplify the exposition or suggest the correct intuition (while possibly minimizing errors and confusion at the same time). )  (*Wikipedia*)

and I agree that your (ab)uses have helped simplify the exposition and are helping suggest the correct intuition.  So let's continue following those suggestions towards the correct intuition.

Now, as to the idea of an iterator: interesting, not bad, maybe a bit un-Pythonic in that usually in NumPy we try to avoid explicit iteration over ndarrays in favour of built-in vectorized operations, but perhaps there might be a case to be made for its use in development as an explicit way which can be asserted to do the same thing as a less obvious `einsum`?  Yes, I can see that.

I've been away on holidays and see that I've still quite a bit to catch up on, but before that I did begin on working to create somewhat similar figures on triangular and quadrilateral meshes, indicating quadrature points, normals to facets, &c., without recourse to side-effects of `asm`.  This was in &lsquo;render ElementTriRT0 field&rsquo; #824, but I think you already saw that towards the end of last year.  

Rather than iterators, that accessed ndarrays of `FacetBasis.global_coordinates()` and `FacetBasis.normals`.  Do we need iterators rather than arrays?  If one really wants to iterate, one can always iterate over an ndarray with [`nditer`](https://numpy.org/devdocs/reference/arrays.nditer.html).  Are there properties or attributes that we would like to iterate over that aren't yet as easily available as `.global_coordinates()` and `.normals`?
 
I'll take a look at #851 and after that I hope to have more concrete suggestions.  It's something that I'm very interested in anyway.  We really want to be able to calculate the jump and flux across an interface between subdomains as in ex26 and that means being able to consistently reorient the constitutive facets #821 or at least know what their orientation is so that it can be accounted for.

**Reply by gatling-nrl** [reply]
There are two #851's... before @kinnala helped me out and after. He helped me get rid of my use of asm inside the basis function. But I think it's worth taking a look at how I did it.

And no worries about the abuse of notation. I wear it like a badge of honor :)  Another idiom that makes me smile when reading a derivation is something along the lines of "through malice of forethought, we're going to set `x=this_thing_that_makes_no_sense`". And of course that creative representation of `x` always works out.

After @kinnala jumped in, I realized that an iterator is not needed. I would have deleted this discussion if possible. The trouble is really that some of the lower levels of the library are a bit cryptic. `mesh.t2f` and `tind` were hard to decipher. I started expanding the docstrings and comments in #851

As a final thought tonight since I'm off to bed, I'm reasonably happy with #851, except that it's giving me wrong answers for toy problems. I'm going to dig into the physics tomorrow and if needed bring in a COMSOL colleague to help me out. Hopefully its something silly, and hopefully by end of tomorrow I can say #851 is good to go.... (as far as I'm concerned).

**Reply by gatling-nrl** [reply]
> We really want to be able to calculate the jump and flux across an interface between subdomains

I'm not sure what "jump" is.. I've seen that word used other places, like in reference to using a list of bases in `asm`. Do you think #851 currently addresses this? As for flux, that's basically my use case, and #851 seems to do that quite well.

**Reply by gdmcbain** [reply]
Imagine that ex26 were extended to solve not just in the core subdomain but also in the annulus. if a Dirichlet condition on temperature were used on the interface in both problems, the jump would be zero, but the fluxes wouldn't match. Unless we somehow managed to specify the temperature at just the right values.

Conversely, if we had specified the flux on the interface as a Neumann condition for the separate subdomains, the temperatures wouldn't match. The difference in temperature at the interface from thw two sides is called the jump. (Note here that i mean the local flux density at each point in the interface, not the integarted flux, or "current" in the electrical analogue.)

Sometimes we don't want the jumo to vanish; e.g. for thermal contact resistance, in which there is a jump proportional to minus the flux.

A specified mismatch in fluxes is also possible. This could corrspond to an interfacial source, generalizing a point source.

See my 2012 slides from the Fourth FreeFem workshop : [Domain decomposition techniques for interfacial discontinuities](https://gdmcbain.github.io/papers/mcbain_12_domain.pdf). This is all really easy in FreeFEM because it's possible to define (n-1)-dimensional bits of the mesh (basically "boundaries" but not necessarily actually on the boundary, really just subsets of facets) that are consistently oriented.

I hadn't conceived of these interfaces as bounding subdomains, although they do in special cases like your isolated island subdomains. I thought of them mostly as separating two subdomains; not always the same thing, e.g. if three subdomains meet at a point (in the plane, or at a curve in space). I have also even had cases (in FreeFEM) in which there was an internal curvilinear (well, polygonal anyway) segment inside a two-dimensional domain. This represented a thin electrically heated sheet embedded in a thermally conducting solid. This kind of topology is easy to specify in Gmsh (I was using the old GEO syntax then, before the new Python API) and easy to represent in Gmsh's MSH formats using lower-dimensional elements. It's then easy to read in to scikit-fem as something a bit like a boundary (even though it doesn't bound anything) but the infrastructure isn't set up for this. The obvious thing is that the facets aren't consistently oriented.

It's not clear to me that #851 is well set up to handle such cases. It might not get in the way, but I would like the orientation part factored out so that I can make sense of the orientation of any continuous collection of facets. This needn't hold up your current effort, but I'm flagging that either I might have to use something else later or refactor it if I return to these interfacial discontinuity problems.

**Reply by kinnala** [reply]
Thanks @gdmcbain for an interesting comment. I think there are many ways to hack `InteriorFacetBasis` to use a consistent normal, however, they won't be very general. Not every collection of facets is continuous so trying to orient the normals by default is not maybe the correct approach?

**Reply by gdmcbain** [reply]
I'm starting to think that that's right, yes. We do see any example of a discontinuous set of facets in the application of `SubdomainFacetBasis` to a disconnected subdomain.

I'm starting to wonder whether what's really wanted for the completion of ex26 is something very like mortar elements but supported on facets of the n-domensional mesh rather than a separate (n-1)-dinensional mesh on an independent hypersurface. Thus this "mortar" would be connecting compatibly meshed subdomains; connecting but possibly also dividing with inhomigeneous jump and flux-mismatch conditions.

I'm wondering whether it would be better to proceed more cautiously, solving examples with the library as is putting all the extra code (identification and orientation of facets, calculations of local jumps and flux-mismatches) in the top level script. Then if it appears that there is something there that is going to be of general interest and that can be reasonably factored into a function, then that can be considered for inclusion in the library. This is how I was proceeding in ex26, e.g. at

https://github.com/kinnala/scikit-fem/blob/b24b078308f89a2fff63882939904f3fd8d7d17f/docs/examples/ex26.py#L34-L37

A possible next step in that example would be to calculate the local flux out of the core to use as a Neumann condition for the annulus, then compute the jump across the interface and iterate, recrossing the Dirichlet condition for the core on the interface until the jump vanishes. Such Dirichlet–Neumann conditions don't always converge, depending on the ratio of thermal conductivities. Better is to write a function that maps core interfacial temperature to interfacial jump and then pass that to a solver—same structure as the Uzawa-CG sklver for preseure–incompressibility in

https://github.com/kinnala/scikit-fem/blob/b24b078308f89a2fff63882939904f3fd8d7d17f/docs/examples/ex30.py#L96-L101

&hellip;or, going further, including everything in the system matrix and solving that, perhaps with preconditioned iterations, as in

https://github.com/kinnala/scikit-fem/blob/b24b078308f89a2fff63882939904f3fd8d7d17f/docs/examples/ex32.py#L139-L149

But first step is orientation. #821 I think we've got a couple of techniques for that now #824 #851, but let's factor these out and clean them up and not abuse `asm`, which was not necessary in #824 though it didn't yet go on to calculate the one-sided fluxes.

One question to think about is what function space the one-sided fluxes and flux-mismatch should be in. It's like a Lagrange multiplier space. For techniques as described above it's most simply the same finite dimension as the jump so that the solving makes sense.

**Reply by gdmcbain** [reply]
A special case of what I'm describing here could restrict the space of core-interfacial temperatures to &reals;; I.e. to be a single undetermined value. This would be associated with the integrated flux. This would be a rational resolution of #751 and maybe the ultimate goal of #851 though I'm not exactly sure what that is.

**Reply by kinnala** [reply]
Technically the normal vector is a finite element function defined by ElementVector(ElementTriSkeletonP0()). However, I'm not happy about how these skeleton elements are implemented and I mean to do something about it.
